### ISSUE TITLE
Mixer path has some silent float-to-double promotions

### ISSUE THREAD

**Issue by kcgen** [dosbox-staging/dosbox-staging#2006]
The team has done some comparisons to show that floating-point processing is faster than doubles (on all platforms and compilers, ranging from 25% to roughly 50% faster). 

There are still some promtions where floats are becoming doubles in the mixer pathway, mostly due to 3rd party code.

It would be nice to eliminate all of these and have a "float-clean" route. 

These extra type conversion warnings can be shown with:

``` shell
export CXX="clang++ -Wconversion -Wsign-conversion -Wdouble-promotion"
touch src/hardware/mixer.cpp
meson setup build
meson compile -C build
```

```
In file included from ../../src/hardware/mixer.cpp:65:
In file included from
../../include/../src/libs/tal-chorus/ChorusEngine.h:27:
In file included from ../../include/../src/libs/tal-chorus/Chorus.h:27:
../../include/../src/libs/tal-chorus/Lfo.h:55:56: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
void setSampleRate(float samplerate_) {samplerate = (samplerate_>0.0) ?
samplerate :
44100.0f;}
^~~~~~~~~~~~
In file included from ../../src/hardware/mixer.cpp:65:
In file included from
../../include/../src/libs/tal-chorus/ChorusEngine.h:27:
../../include/../src/libs/tal-chorus/Chorus.h:83:28: warning: implicit
conversion changes signedness: 'int' to 'unsigned long'
[-Wsign-conversion]
delayLineStart(new float[delayLineLength]),
~~~       ^~~~~~~~~~~~~~~
../../include/../src/libs/tal-chorus/Chorus.h:140:16: warning: implicit
conversion from 'int' to 'float' may lose precision
[-Wimplicit-int-float-conversion]
frac= offset-(int)floorf(offset);
~^~~~~~~~~~~~~~~~~~~
In file included from ../../src/hardware/mixer.cpp:65:
../../include/../src/libs/tal-chorus/ChorusEngine.h:72:32: warning:
declaration shadows a field of 'ChorusEngine' [-Wshadow]
void setEnablesChorus(bool isChorus1Enabled, bool
isChorus2Enabled)
^
../../include/../src/libs/tal-chorus/ChorusEngine.h:45:10: note: previous
declaration is here
bool isChorus1Enabled = false;
^
../../include/../src/libs/tal-chorus/ChorusEngine.h:72:55: warning:
declaration shadows a field of 'ChorusEngine' [-Wshadow]
void setEnablesChorus(bool isChorus1Enabled, bool
isChorus2Enabled)
^
../../include/../src/libs/tal-chorus/ChorusEngine.h:46:10: note: previous
declaration is here
bool isChorus2Enabled = false;
^
../../src/hardware/mixer.cpp:141:25: warning: implicit conversion
increases floating-point precision: 'const float' to 'double'
[-Wdouble-promotion]
f.setup(sample_rate, highpass_freq);
~~~~~              ^~~~~~~~~~~~~
../../src/hardware/mixer.cpp:257:41: warning: implicit conversion
increases floating-point precision: 'float const' to 'double'
[-Wdouble-promotion]
if (std::isfinite(cf) && cf >= 0.0 && cf <= 100.0)
{
^~ ~~
../../src/hardware/mixer.cpp:257:28: warning: implicit conversion
increases floating-point precision: 'float const' to 'double'
[-Wdouble-promotion]
if (std::isfinite(cf) && cf >= 0.0 && cf <= 100.0)
{
^~ ~~
../../src/hardware/mixer.cpp:453:29: warning: implicit conversion loses
integer precision: '__int_type' (aka 'int') to 'uint16_t' (aka 'unsigned
short') [-Wimplicit-int-conversion]
mixer.compressor.Configure(mixer.sample_rate,
~~~~~~~~~ ^~~~~~~~~~~~~~~~~
../../src/hardware/mixer.cpp:701:35: warning: implicit conversion changes
signedness: 'const __int_type' (aka 'const int') to 'spx_uint32_t' (aka
'unsigned int') [-Wsign-conversion]
num_channels, in_rate, out_rate, quality,
nullptr);
^~~~~~~~
../../src/hardware/mixer.cpp:701:26: warning: implicit conversion changes
signedness: 'const int' to 'spx_uint32_t' (aka 'unsigned int')
[-Wsign-conversion]
num_channels, in_rate, out_rate, quality,
nullptr);
^~~~~~~
../../src/hardware/mixer.cpp:703:60: warning: implicit conversion changes
signedness: 'const __int_type' (aka 'const int') to 'spx_uint32_t' (aka
'unsigned int') [-Wsign-conversion]
speex_resampler_set_rate(speex_resampler.state, in_rate,
out_rate);
~~~~~~~~~~~~~~~~~~~~~~~~                                 ^~~~~~~~
../../src/hardware/mixer.cpp:703:51: warning: implicit conversion changes
signedness: 'const int' to 'spx_uint32_t' (aka 'unsigned int')
[-Wsign-conversion]
speex_resampler_set_rate(speex_resampler.state, in_rate,
out_rate);
~~~~~~~~~~~~~~~~~~~~~~~~                        ^~~~~~~
../../src/hardware/mixer.cpp:835:21: warning: implicit conversion changes
signedness: 'int' to 'std::size_t' (aka 'unsigned long')
[-Wsign-conversion]
if (prev_frame[ch] > f)
~~~~~~~~~~ ^~
../../src/hardware/mixer.cpp:836:18: warning: implicit conversion changes
signedness: 'int' to 'std::size_t' (aka 'unsigned long')
[-Wsign-conversion]
next_frame[ch] = prev_frame[ch] -
f;
~~~~~~~~~~ ^~
../../src/hardware/mixer.cpp:836:35: warning: implicit conversion changes
signedness: 'int' to 'std::size_t' (aka 'unsigned long')
[-Wsign-conversion]
next_frame[ch] = prev_frame[ch] -
f;
~~~~~~~~~~ ^~
../../src/hardware/mixer.cpp:837:26: warning: implicit conversion changes
signedness: 'int' to 'std::size_t' (aka 'unsigned long')
[-Wsign-conversion]
else if (prev_frame[ch] <
-f)
~~~~~~~~~~ ^~
../../src/hardware/mixer.cpp:838:18: warning: implicit conversion changes
signedness: 'int' to 'std::size_t' (aka 'unsigned long')
[-Wsign-conversion]
next_frame[ch] = prev_frame[0] +
f;
~~~~~~~~~~ ^~
../../src/hardware/mixer.cpp:840:18: warning: implicit conversion changes
signedness: 'int' to 'std::size_t' (aka 'unsigned long')
[-Wsign-conversion]
next_frame[ch] = 0.0f;
~~~~~~~~~~ ^~
../../src/hardware/mixer.cpp:989:34: warning: implicit conversion loses
integer precision: 'int' to 'uint16_t' (aka 'unsigned short')
[-Wimplicit-int-conversion]
ConfigureLowPassFilter(order, cutoff_freq_hz);
~~~~~~~~~~~~~~~~~~~~~~        ^~~~~~~~~~~~~~
../../src/hardware/mixer.cpp:989:27: warning: implicit conversion loses
integer precision: 'int' to 'uint8_t' (aka 'unsigned char')
[-Wimplicit-int-conversion]
ConfigureLowPassFilter(order, cutoff_freq_hz);
~~~~~~~~~~~~~~~~~~~~~~ ^~~~~
../../src/hardware/mixer.cpp:992:35: warning: implicit conversion loses
integer precision: 'int' to 'uint16_t' (aka 'unsigned short')
[-Wimplicit-int-conversion]
ConfigureHighPassFilter(order, cutoff_freq_hz);
~~~~~~~~~~~~~~~~~~~~~~~        ^~~~~~~~~~~~~~
../../src/hardware/mixer.cpp:992:28: warning: implicit conversion loses
integer precision: 'int' to 'uint8_t' (aka 'unsigned char')
[-Wimplicit-int-conversion]
ConfigureHighPassFilter(order, cutoff_freq_hz);
~~~~~~~~~~~~~~~~~~~~~~~ ^~~~~
../../src/hardware/mixer.cpp:1006:42: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
const auto filter_type        =
parts[i++];
~~~~~ ~^~
../../src/hardware/mixer.cpp:1007:42: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
const auto filter_order       =
parts[i++];
~~~~~ ~^~
../../src/hardware/mixer.cpp:1008:42: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
const auto filter_cutoff_freq =
parts[i++];
~~~~~ ~^~
../../src/hardware/mixer.cpp:1014:43: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
const auto filter1_type        =
parts[i++];
~~~~~ ~^~
../../src/hardware/mixer.cpp:1015:43: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
const auto filter1_order       =
parts[i++];
~~~~~ ~^~
../../src/hardware/mixer.cpp:1016:43: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
const auto filter1_cutoff_freq =
parts[i++];
~~~~~ ~^~
../../src/hardware/mixer.cpp:1018:43: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
const auto filter2_type        =
parts[i++];
~~~~~ ~^~
../../src/hardware/mixer.cpp:1019:43: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
const auto filter2_order       =
parts[i++];
~~~~~ ~^~
../../src/hardware/mixer.cpp:1020:43: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
const auto filter2_cutoff_freq =
parts[i++];
~~~~~ ~^~
../../src/hardware/mixer.cpp:1057:66: warning: implicit conversion from
'__int_type' (aka 'int') to 'float' may lose precision
[-Wimplicit-int-float-conversion]
mixer.sample_rate.load(),
~~~~~~~~~~~~~~~~~~^~~~~~
../../src/hardware/mixer.cpp:1124:56: warning: implicit conversion
increases floating-point precision: 'float const' to 'double'
[-Wdouble-promotion]
reverb.send_gain =
static_cast<float>(decibel_to_gain(level_db));
~~~~~~~~~~~~~~~ ^~~~~~~~
../../src/hardware/mixer.cpp:1161:56: warning: implicit conversion
increases floating-point precision: 'float const' to 'double'
[-Wdouble-promotion]
chorus.send_gain =
static_cast<float>(decibel_to_gain(level_db));
~~~~~~~~~~~~~~~ ^~~~~~~~
../../src/hardware/mixer.cpp:1817:11: warning: implicit conversion changes
signedness: 'int' to 'std::size_t' (aka 'unsigned long')
[-Wsign-conversion]
frame[ch] =
mixer.reverb.highpass_filter[ch].filter(
~~~~~ ^~
../../src/hardware/mixer.cpp:1818:19: warning: implicit conversion changes
signedness: 'int' to 'std::size_t' (aka 'unsigned long')
[-Wsign-conversion]
frame[ch]);
~~~~~ ^~
../../src/hardware/mixer.cpp:1817:46: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
frame[ch] =
mixer.reverb.highpass_filter[ch].filter(
~~~~~                        ^~
../../src/hardware/mixer.cpp:1863:20: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
mixer.work[pos][ch] =
mixer.highpass_filter[ch].filter(
~~~~~           ^~
../../src/hardware/mixer.cpp:1864:28: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
mixer.work[pos][ch]);
~~~~~           ^~
../../src/hardware/mixer.cpp:1863:48: warning: implicit conversion changes
signedness: 'int' to 'size_type' (aka 'unsigned long')
[-Wsign-conversion]
mixer.work[pos][ch] =
mixer.highpass_filter[ch].filter(
~~~~~                 ^~
../../src/hardware/mixer.cpp:1888:8: warning: declaration shadows a local
variable [-Wshadow]
auto pos = start_pos;
^
../../src/hardware/mixer.cpp:1859:7: note: previous declaration is
here
auto pos = start_pos;
^
../../src/hardware/mixer.cpp:1903:19: warning: implicit conversion changes
signedness: '__int_type' (aka 'int') to 'uint32_t' (aka 'unsigned int')
[-Wsign-conversion]
CAPTURE_AddWave(mixer.sample_rate,
~~~~~~~~~~~~~~~ ^~~~~~~~~~~~~~~~~
../../src/hardware/mixer.cpp:1961:20: warning: implicit conversion changes
signedness: 'int' to 'size_t' (aka 'unsigned long')
[-Wsign-conversion]
memset(stream, 0, len);
~~~~~~            ^~~
../../src/hardware/mixer.cpp:2326:30: warning: implicit conversion
increases floating-point precision: 'float const' to 'double'
[-Wdouble-promotion]
decibel_to_gain(vol));
~~~~~~~~~~~~~~~ ^~~
../../src/hardware/mixer.cpp:2367:36: warning: implicit conversion
increases floating-point precision: 'const float' to 'double'
[-Wdouble-promotion]
gain_to_decibel(volume.right),
~~~~~~~~~~~~~~~ ~~~~~~~^~~~~
../../src/hardware/mixer.cpp:2366:36: warning: implicit conversion
increases floating-point precision: 'const float' to 'double'
[-Wdouble-promotion]
gain_to_decibel(volume.left),
~~~~~~~~~~~~~~~ ~~~~~~~^~~~
../../src/hardware/mixer.cpp:2365:26: warning: implicit conversion
increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
volume.right * 100.0f,
~~~~~~~~~~~~~^~~~~~~~
../../src/hardware/mixer.cpp:2364:25: warning: implicit conversion
increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
volume.left * 100.0f,
~~~~~~~~~~~~^~~~~~~~
In file included from ../../src/hardware/mixer.cpp:64:
../../include/../src/libs/mverb/MVerb.h:888:61: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
f = static_cast<T>(2.0 * sin(M_PI * frequency /
sampleRate));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:845:13: note: in instantiation of
member function 'StateVariable<float, 4>::UpdateCoefficient' requested
here
UpdateCoefficient();
^
../../include/../src/libs/mverb/MVerb.h:818:13: note: in instantiation of
member function 'StateVariable<float, 4>::SetSampleRate' requested
here
SetSampleRate(44100.);
^
../../include/../src/libs/mverb/MVerb.h:37:46: note: in instantiation of
member function 'StateVariable<float, 4>::StateVariable' requested
here
StateVariable<T,4> bandwidthFilter[2] =
{};
^
../../src/hardware/mixer.cpp:105:15: note: in instantiation of member
function 'MVerb<float>::MVerb' requested here
EmVerb mverb = {};
^
In file included from ../../src/hardware/mixer.cpp:64:
../../include/../src/libs/mverb/MVerb.h:888:49: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
f = static_cast<T>(2.0 * sin(M_PI * frequency /
sampleRate));
~ ^~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:225:56: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpass[0].SetLength(static_cast<int>(0.0048 *
SampleRate));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:101:9: note: in instantiation of
member function 'MVerb<float>::reset' requested here
reset();
^
../../src/hardware/mixer.cpp:105:15: note: in instantiation of member
function 'MVerb<float>::MVerb' requested here
EmVerb mverb = {};
^
In file included from ../../src/hardware/mixer.cpp:64:
../../include/../src/libs/mverb/MVerb.h:226:56: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpass[1].SetLength(static_cast<int>(0.0036 *
SampleRate));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:227:56: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpass[2].SetLength(static_cast<int>(0.0127 *
SampleRate));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:228:56: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpass[3].SetLength(static_cast<int>(0.0093 *
SampleRate));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:237:75: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[0].SetLength(static_cast<int>(0.020 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:237:62: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[0].SetLength(static_cast<int>(0.020 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:238:75: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetLength(static_cast<int>(0.060 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:238:62: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetLength(static_cast<int>(0.060 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:239:75: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[2].SetLength(static_cast<int>(0.030 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:239:62: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[2].SetLength(static_cast<int>(0.030 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:240:75: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetLength(static_cast<int>(0.089 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:240:62: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetLength(static_cast<int>(0.089 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:246:122: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetIndex(0, static_cast<int>(0.006 * SampleRate * Size),
static_cast<int>(0.041 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:246:109: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetIndex(0, static_cast<int>(0.006 * SampleRate * Size),
static_cast<int>(0.041 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:246:77: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetIndex(0, static_cast<int>(0.006 * SampleRate * Size),
static_cast<int>(0.041 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:246:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetIndex(0, static_cast<int>(0.006 * SampleRate * Size),
static_cast<int>(0.041 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:248:122: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetIndex(0, static_cast<int>(0.031 * SampleRate * Size),
static_cast<int>(0.011 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:248:109: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetIndex(0, static_cast<int>(0.031 * SampleRate * Size),
static_cast<int>(0.011 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:248:77: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetIndex(0, static_cast<int>(0.031 * SampleRate * Size),
static_cast<int>(0.011 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:248:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetIndex(0, static_cast<int>(0.031 * SampleRate * Size),
static_cast<int>(0.011 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:253:75: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetLength(static_cast<int>(0.15 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:253:62: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetLength(static_cast<int>(0.15 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:254:75: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetLength(static_cast<int>(0.12 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:254:62: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetLength(static_cast<int>(0.12 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:255:75: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetLength(static_cast<int>(0.14 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:255:62: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetLength(static_cast<int>(0.14 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:256:75: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetLength(static_cast<int>(0.11 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:256:62: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetLength(static_cast<int>(0.11 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:257:168: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:257:155: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:257:123: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:257:110: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:257:78: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:257:65: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:258:123: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetIndex(0, static_cast<int>(0.036 * SampleRate *
Size), static_cast<int>(0.089 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:258:110: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetIndex(0, static_cast<int>(0.036 * SampleRate *
Size), static_cast<int>(0.089 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:258:78: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetIndex(0, static_cast<int>(0.036 * SampleRate *
Size), static_cast<int>(0.089 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:258:65: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetIndex(0, static_cast<int>(0.036 * SampleRate *
Size), static_cast<int>(0.089 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:259:124: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetIndex(0, static_cast<int>(0.0089 * SampleRate *
Size), static_cast<int>(0.099 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:259:111: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetIndex(0, static_cast<int>(0.0089 * SampleRate *
Size), static_cast<int>(0.099 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:259:79: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetIndex(0, static_cast<int>(0.0089 * SampleRate *
Size), static_cast<int>(0.099 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:259:66: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetIndex(0, static_cast<int>(0.0089 * SampleRate *
Size), static_cast<int>(0.099 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:260:124: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.0041 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:260:111: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.0041 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:260:78: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.0041 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:260:65: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.0041 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:263:73: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[0].SetLength(static_cast<int>(0.089 *
SampleRate));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:264:271: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[0].SetIndex(0, static_cast<int>(0.0199 *
SampleRate), static_cast<int>(0.0219 * SampleRate),
static_cast<int>(0.0354 * SampleRate), static_cast<int>(0.0389 *
SampleRate), static_cast<int>(0.0414 * SampleRate),
static_cast<int>(0.0692 * SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:264:232: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[0].SetIndex(0, static_cast<int>(0.0199 *
SampleRate), static_cast<int>(0.0219 * SampleRate),
static_cast<int>(0.0354 * SampleRate), static_cast<int>(0.0389 *
SampleRate), static_cast<int>(0.0414 * SampleRate),
static_cast<int>(0.0692 * SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:264:193: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[0].SetIndex(0, static_cast<int>(0.0199 *
SampleRate), static_cast<int>(0.0219 * SampleRate),
static_cast<int>(0.0354 * SampleRate), static_cast<int>(0.0389 *
SampleRate), static_cast<int>(0.0414 * SampleRate),
static_cast<int>(0.0692 * SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:264:154: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[0].SetIndex(0, static_cast<int>(0.0199 *
SampleRate), static_cast<int>(0.0219 * SampleRate),
static_cast<int>(0.0354 * SampleRate), static_cast<int>(0.0389 *
SampleRate), static_cast<int>(0.0414 * SampleRate),
static_cast<int>(0.0692 * SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:264:115: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[0].SetIndex(0, static_cast<int>(0.0199 *
SampleRate), static_cast<int>(0.0219 * SampleRate),
static_cast<int>(0.0354 * SampleRate), static_cast<int>(0.0389 *
SampleRate), static_cast<int>(0.0414 * SampleRate),
static_cast<int>(0.0692 * SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:264:76: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[0].SetIndex(0, static_cast<int>(0.0199 *
SampleRate), static_cast<int>(0.0219 * SampleRate),
static_cast<int>(0.0354 * SampleRate), static_cast<int>(0.0389 *
SampleRate), static_cast<int>(0.0414 * SampleRate),
static_cast<int>(0.0692 * SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:265:73: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[1].SetLength(static_cast<int>(0.069 *
SampleRate));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:266:270: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[1].SetIndex(0, static_cast<int>(0.0099 *
SampleRate), static_cast<int>(0.011 * SampleRate), static_cast<int>(0.0182
* SampleRate), static_cast<int>(0.0189 * SampleRate),
static_cast<int>(0.0213 * SampleRate), static_cast<int>(0.0431 *
SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:266:231: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[1].SetIndex(0, static_cast<int>(0.0099 *
SampleRate), static_cast<int>(0.011 * SampleRate), static_cast<int>(0.0182
* SampleRate), static_cast<int>(0.0189 * SampleRate),
static_cast<int>(0.0213 * SampleRate), static_cast<int>(0.0431 *
SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:266:192: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[1].SetIndex(0, static_cast<int>(0.0099 *
SampleRate), static_cast<int>(0.011 * SampleRate), static_cast<int>(0.0182
* SampleRate), static_cast<int>(0.0189 * SampleRate),
static_cast<int>(0.0213 * SampleRate), static_cast<int>(0.0431 *
SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:266:153: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[1].SetIndex(0, static_cast<int>(0.0099 *
SampleRate), static_cast<int>(0.011 * SampleRate), static_cast<int>(0.0182
* SampleRate), static_cast<int>(0.0189 * SampleRate),
static_cast<int>(0.0213 * SampleRate), static_cast<int>(0.0431 *
SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:266:114: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[1].SetIndex(0, static_cast<int>(0.0099 *
SampleRate), static_cast<int>(0.011 * SampleRate), static_cast<int>(0.0182
* SampleRate), static_cast<int>(0.0189 * SampleRate),
static_cast<int>(0.0213 * SampleRate), static_cast<int>(0.0431 *
SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:266:76: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
earlyReflectionsDelayLine[1].SetIndex(0, static_cast<int>(0.0099 *
SampleRate), static_cast<int>(0.011 * SampleRate), static_cast<int>(0.0182
* SampleRate), static_cast<int>(0.0189 * SampleRate),
static_cast<int>(0.0213 * SampleRate), static_cast<int>(0.0431 *
SampleRate),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:272:55: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
DampingFreq = static_cast<T>(1. -
value);
~ ^~~~~
../../src/hardware/mixer.cpp:128:9: note: in instantiation of member
function 'MVerb<float>::setParameter' requested here
mverb.setParameter(EmVerb::PREDELAY, predelay);
^
In file included from ../../src/hardware/mixer.cpp:64:
../../include/../src/libs/mverb/MVerb.h:284:51: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
Size = static_cast<T>((0.95 * value) +
0.05);
~ ^~~~~
../../include/../src/libs/mverb/MVerb.h:289:87: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[0].SetLength(static_cast<int>(0.020 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:289:74: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[0].SetLength(static_cast<int>(0.020 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:290:87: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetLength(static_cast<int>(0.060 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:290:74: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetLength(static_cast<int>(0.060 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:291:87: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[2].SetLength(static_cast<int>(0.030 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:291:74: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[2].SetLength(static_cast<int>(0.030 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:292:87: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetLength(static_cast<int>(0.089 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:292:74: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetLength(static_cast<int>(0.089 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:293:134: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetIndex(0, static_cast<int>(0.006 * SampleRate * Size),
static_cast<int>(0.041 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:293:121: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetIndex(0, static_cast<int>(0.006 * SampleRate * Size),
static_cast<int>(0.041 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:293:89: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetIndex(0, static_cast<int>(0.006 * SampleRate * Size),
static_cast<int>(0.041 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:293:76: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[1].SetIndex(0, static_cast<int>(0.006 * SampleRate * Size),
static_cast<int>(0.041 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:294:134: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetIndex(0, static_cast<int>(0.031 * SampleRate * Size),
static_cast<int>(0.011 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:294:121: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetIndex(0, static_cast<int>(0.031 * SampleRate * Size),
static_cast<int>(0.011 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:294:89: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetIndex(0, static_cast<int>(0.031 * SampleRate * Size),
static_cast<int>(0.011 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:294:76: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
allpassFourTap[3].SetIndex(0, static_cast<int>(0.031 * SampleRate * Size),
static_cast<int>(0.011 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:299:87: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetLength(static_cast<int>(0.15 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:299:74: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetLength(static_cast<int>(0.15 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:300:87: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetLength(static_cast<int>(0.12 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:300:74: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetLength(static_cast<int>(0.12 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:301:87: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetLength(static_cast<int>(0.14 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:301:74: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetLength(static_cast<int>(0.14 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:302:87: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetLength(static_cast<int>(0.11 * SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:302:74: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetLength(static_cast<int>(0.11 * SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:303:180: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:303:167: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:303:135: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:303:122: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:303:90: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:303:77: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[0].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.011 * SampleRate * Size), static_cast<int>(0.121
* SampleRate *
Size));
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:304:135: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetIndex(0, static_cast<int>(0.036 * SampleRate *
Size), static_cast<int>(0.089 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:304:122: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetIndex(0, static_cast<int>(0.036 * SampleRate *
Size), static_cast<int>(0.089 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:304:90: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetIndex(0, static_cast<int>(0.036 * SampleRate *
Size), static_cast<int>(0.089 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:304:77: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[1].SetIndex(0, static_cast<int>(0.036 * SampleRate *
Size), static_cast<int>(0.089 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:305:136: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetIndex(0, static_cast<int>(0.0089 * SampleRate *
Size), static_cast<int>(0.099 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:305:123: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetIndex(0, static_cast<int>(0.0089 * SampleRate *
Size), static_cast<int>(0.099 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:305:91: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetIndex(0, static_cast<int>(0.0089 * SampleRate *
Size), static_cast<int>(0.099 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:305:78: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[2].SetIndex(0, static_cast<int>(0.0089 * SampleRate *
Size), static_cast<int>(0.099 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:306:136: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.0041 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:306:123: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.0041 * SampleRate * Size),
0);
~ ^~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:306:90: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.0041 * SampleRate * Size),
0);
~ ^~~~
../../include/../src/libs/mverb/MVerb.h:306:77: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
staticDelayLine[3].SetIndex(0, static_cast<int>(0.067 * SampleRate *
Size), static_cast<int>(0.0041 * SampleRate * Size),
0);
~ ^~~~~~~~~~
In file included from ../../src/hardware/mixer.cpp:24:
In file included from ../../include/mixer.h:42:
In file included from ../../subprojects/iir1-1.9.3/Iir.h:45:
../../subprojects/iir1-1.9.3/iir/Cascade.h:164:30: warning: implicit
conversion increases floating-point precision: 'const float' to 'double'
[-Wdouble-promotion]
double out = in;
~~~   ^~
../../src/hardware/mixer.cpp:1559:41: note: in instantiation of function
template specialization 'Iir::CascadeStages<8,
Iir::DirectFormII>::filter<float>' requested here
frame.left =
filters.highpass.hpf[0].filter(frame.left);
^
../../src/hardware/mixer.cpp:1654:2: note: in instantiation of function
template specialization 'MixerChannel::AddSamples<unsigned char, false,
false, true>' requested here
AddSamples<uint8_t, false, false, true>(len, data);
^
In file included from ../../src/hardware/mixer.cpp:24:
In file included from ../../include/mixer.h:42:
In file included from ../../subprojects/iir1-1.9.3/Iir.h:45:
../../subprojects/iir1-1.9.3/iir/Cascade.h:164:30: warning: implicit
conversion increases floating-point precision: 'const float' to 'double'
[-Wdouble-promotion]
double out = in;
~~~   ^~
../../src/hardware/mixer.cpp:1817:50: note: in instantiation of function
template specialization 'Iir::CascadeStages<1,
Iir::DirectFormII>::filter<float>' requested
here
frame[ch] =
mixer.reverb.highpass_filter[ch].filter(
^
In file included from ../../src/hardware/mixer.cpp:64:
../../include/../src/libs/mverb/MVerb.h:121:85: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T DecayDelta = static_cast<T>((((0.7995f * Decay) + 0.005) - DecaySmooth)
*
OneOverSampleFrames);
~ ^~~~~~~~~~~~~~~~~~~
../../src/hardware/mixer.cpp:1831:23: note: in instantiation of member
function 'MVerb<float>::process' requested here
mixer.reverb.mverb.process(in, out,
frames);
^
In file included from ../../src/hardware/mixer.cpp:64:
../../include/../src/libs/mverb/MVerb.h:121:70: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T DecayDelta = static_cast<T>((((0.7995f * Decay) + 0.005) - DecaySmooth)
*
OneOverSampleFrames);
~ ^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:121:50: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T DecayDelta = static_cast<T>((((0.7995f * Decay) + 0.005) - DecaySmooth)
* OneOverSampleFrames);
~~~~~~~~^~~~~~~  ~
../../include/../src/libs/mverb/MVerb.h:122:92: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T DensityDelta = static_cast<T>((((0.7995f * Density1) + 0.005) -
DensitySmooth) *
OneOverSampleFrames);
~ ^~~~~~~~~~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:122:75: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T DensityDelta = static_cast<T>((((0.7995f * Density1) + 0.005) -
DensitySmooth) *
OneOverSampleFrames);
~ ^~~~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:122:52: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T DensityDelta = static_cast<T>((((0.7995f * Density1) + 0.005) -
DensitySmooth) *
OneOverSampleFrames);
~~~~~~~~^~~~~~~~~~  ~
../../include/../src/libs/mverb/MVerb.h:143:39: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
Density2 = static_cast<T>(DecaySmooth +
0.15);
^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:144:17: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
if (Density2 > 0.5)
^~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:146:17: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
if (Density2 < 0.25)
^~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:161:59: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ ( bandwidthLeft * 0.4 + bandwidthRight * 0.2 ) *
0.5);
^~~~~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:161:37: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ ( bandwidthLeft * 0.4 + bandwidthRight * 0.2 ) *
0.5);
^~~~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:160:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[0].GetIndex(7) *
0.1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:159:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[0].GetIndex(6) *
0.1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:158:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[0].GetIndex(5) *
0.3
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:157:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[0].GetIndex(4) *
0.3
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:156:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[0].GetIndex(3) *
0.4
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:155:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[0].GetIndex(2) *
0.6
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:154:50: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T earlyReflectionsL =
static_cast<T>(earlyReflectionsDelayLine[0](static_cast<T>(bandwidthLeft *
0.5 + bandwidthRight *
0.3))

^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:154:116: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T earlyReflectionsL =
static_cast<T>(earlyReflectionsDelayLine[0](static_cast<T>(bandwidthLeft *
0.5 + bandwidthRight *
0.3))
^~~~~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:154:94: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T earlyReflectionsL =
static_cast<T>(earlyReflectionsDelayLine[0](static_cast<T>(bandwidthLeft *
0.5 + bandwidthRight *
0.3))
^~~~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:169:59: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ ( bandwidthLeft * 0.2 + bandwidthRight * 0.4 ) *
0.5);
^~~~~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:169:37: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ ( bandwidthLeft * 0.2 + bandwidthRight * 0.4 ) *
0.5);
^~~~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:168:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[1].GetIndex(7) *
0.1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:167:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[1].GetIndex(6) *
0.1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:166:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[1].GetIndex(5) *
0.3
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:165:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[1].GetIndex(4) *
0.3
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:164:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[1].GetIndex(3) *
0.4
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:163:64: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+ earlyReflectionsDelayLine[1].GetIndex(2) *
0.6
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:162:50: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T earlyReflectionsR =
static_cast<T>(earlyReflectionsDelayLine[1](static_cast<T>(bandwidthLeft *
0.3 + bandwidthRight *
0.5))

^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:162:116: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T earlyReflectionsR =
static_cast<T>(earlyReflectionsDelayLine[1](static_cast<T>(bandwidthLeft *
0.3 + bandwidthRight *
0.5))
^~~~~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:162:94: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T earlyReflectionsR =
static_cast<T>(earlyReflectionsDelayLine[1](static_cast<T>(bandwidthLeft *
0.3 + bandwidthRight *
0.5))
^~~~~~~~~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:192:54: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
-(0.6*staticDelayLine[1].GetIndex(1)));
~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:191:53: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
-(0.6*allpassFourTap[1].GetIndex(1))
~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:190:54: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
-(0.6*staticDelayLine[0].GetIndex(1))
~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:189:54: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+(0.6*staticDelayLine[3].GetIndex(1))
~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:188:53: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
-(0.6*allpassFourTap[3].GetIndex(1))
~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:187:54: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+(0.6*staticDelayLine[2].GetIndex(2))
~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:186:69: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T accumulatorL =
static_cast<T>((0.6*staticDelayLine[2].GetIndex(1))
~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:199:54: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
-(0.6*staticDelayLine[3].GetIndex(2)));
~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:198:53: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
-(0.6*allpassFourTap[3].GetIndex(2))
~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:197:54: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
-(0.6*staticDelayLine[2].GetIndex(3))
~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:196:54: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+(0.6*staticDelayLine[1].GetIndex(2))
~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:195:53: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
-(0.6*allpassFourTap[1].GetIndex(2))
~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:194:54: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
+(0.6*staticDelayLine[0].GetIndex(3))
~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:193:69: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
T accumulatorR =
static_cast<T>((0.6*staticDelayLine[0].GetIndex(2))
~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~
../../include/../src/libs/mverb/MVerb.h:829:41: warning: implicit
conversion increases floating-point precision: 'float' to 'double'
[-Wdouble-promotion]
low += static_cast<T>(f * band +
1e-25);
~~^~~~~~ ~
../../include/../src/libs/mverb/MVerb.h:152:31: note: in instantiation of
member function 'StateVariable<float, 4>::operator()' requested
here
T bandwidthLeft = bandwidthFilter[0](left) ;
^
../../src/hardware/mixer.cpp:1831:23: note: in instantiation of member
function 'MVerb<float>::process' requested here
mixer.reverb.mverb.process(in, out,
frames);
^
```

**Comment by Burrito78** [comment]
@kcgen did @johnnovak s audio pipeline improvements/rewrite fix these cases?

**Comment by kcgen** [comment]
Not yet, we're still mostly in double-land (when writing new code, it makes sense to be float-clean, but mopping all this other stuff is a lot of work after-the-fact -- so this is low priority, and a good ticket for newcomers to help with).

For MVerb, the changes should land upstream first, so I've opened that ticket: https://github.com/martineastwood/mverb/issues/11

**Comment by johnnovak** [comment]
Careful with this @kcgen. 3rd party audio DSP code might use doubles on purpose. I'm not a DSP expert but know that double precision is necessary for numerical stability in some DSP algorithms.

Not saying it's necessarily the case here, but the filters, reverb, etc. might blow up in interesting ways if they were designed for doubles and now we start using floats. Just tread carefully and test extensively, that's all I'm saying :)

**Comment by kcgen** [comment]
Yup, I'm not aware of the inner working of the algorithm; if they loop tightly and can accumulate precision loss greater than a whole 24bit sample value (or some threshold considered bad by the industry).

Fortunately the warnings catch only unexpected promotion, and won't warn about deliberate casting to double -- so both aspects would help clarify the intent in the code. (I think most of it's just sloppy use of consts, like `var * 6.0` .. oops, that 6.0 just promoted var to a double.

**Comment by johnnovak** [comment]
Yeah, that makes sense. I guess we can just try to use all floats and see.

One area where using doubles is a non-issue is calculating various coefficients at init, so only once based on sample rate, etc. Or when certain parameters change, which should happen rarely, not per sample-frame. So those should continue to use doubles, just in case. The then per-sample-frame calculation (the actual processing) could use floats and we can hope for the best. Then some plugins/effects just _need_ doubles internally.

But if we're not changing the internals, what I wrote above is mostly a moot point. I guess I'm just making this comment to raise awareness on DSP issues 😄

**Comment by johnnovak** [comment]
Probably a non-issue now, I will take a look next time I do work on the mixer.

**Comment by johnnovak** [comment]
Are you intending to work on this in the 0.83.x timeframe @interloper98? If not, I'll pick it up--I just want to avoid us doing the same work twice 😄 

Maybe a non-issue now, btw.

**Comment by interloper98** [comment]
Wasn't in my queue! Go for it 👍

**Comment by johnnovak** [comment]
@weirddan455 How does this relate to your recent warning reporting improvements?

**Comment by weirddan455** [comment]
> [@weirddan455](https://github.com/weirddan455) How does this relate to your recent warning reporting improvements?

It doesn't really. None of these warnings are turned on by default. `-Wconversion` is turned on by `CHECK_NARROWING()` in specific files which normally implies `-Wsign-conversion` as well but sign conversion is disabled because it's too noisy even in the `CHECK_NARROWING()` files. `-Wdouble-promotion` has never been turned on.

My recent PR didn't change change anything with regards to these specific warnings mentioned.

**Comment by johnnovak** [comment]
> My recent PR didn't change change anything with regards to these specific warnings mentioned.

Cheers, leaving it open then. We'll maybe address this one day.
